sound/soc/sof/intel/hda.c

```diff
@@ -958,6 +958,31 @@ int hda_dsp_remove(struct snd_sof_dev *sdev)
 	return 0;
 }
 
+int hda_dsp_shutdown(struct snd_sof_dev *sdev)
+{
+	const struct sof_dsp_power_state target_state = {
+		.state = SOF_DSP_PM_D0,
+		.substate = SOF_HDA_DSP_PM_D0I0,
+	};
+	int ret = 0;
+
+	/*
+	 * turn DSP back to D0 from D0I3 before shutting down it,
+	 * to avoid booting fail at some platforms from D0i3 status.
+	 */
+	if (sdev->dsp_power_state.state == SOF_DSP_PM_D0 &&
+	    sdev->dsp_power_state.substate == SOF_HDA_DSP_PM_D0I3) {
+		/* ensure the DSP is in D0 before shutting it down */
```

Comment:
This seems sufficient as a comment.

Reply:
Maybe dumb question. But why DSP needs to be in D0 before shutting down? Are there any commands sent from Linux driver to DSP at this point and DSP needs to be up to process them?

Reply:
@dbaluta This is a hw/fw workaround effectively for Intel systems. With soft reset, some hw state is not reset by hw, we have existing systems where the state is not reset by BIOS, and it cannot be done from driver directly, so we end up with unable to bring the device up after the subsequent system boot after soft reset.

Reply:
@kv2019i thanks! It makes sense.